lib/active_model_serializers/adapter/json_api.rb

```diff
@@ -0,0 +1,416 @@
+# TODO: implement!
+# TODO: use uri_template in link generation?
+# TODO: validate against a JSON schema document?
+module ActiveModelSerializers
+  module Adapter
+    class JsonApi
+
+      def initialize(serializable_resource)
+        @serializable_resource = serializable_resource
+      end
+
+      def as_json
+        if serializable_resource.success?
+          success
+        else
+          errors
+        end
+      end
+
+      # definition:
+      #   ☐ toplevel_data (required)
+      #   ☐ toplevel_included
+      #   ☑ toplevel_meta
+      #   ☑ toplevel_links
+      #   ☑ toplevel_jsonapi
+      def success
+        {
+          data: toplevel_data,
+          included: toplevel_included,
+          meta: toplevel_meta,
+          links: toplevel_links,
+          jsonapi: toplevel_jsonapi
+        }.reject! {|_,v| v.nil? }
+      end
+
+      # definition:
+      #   ☐ toplevel_errors (required)
+      #   ☑ toplevel_meta
+      #   ☑ toplevel_jsonapi
+      def failure
+        {
+          errors: toplevel_errors,
+          meta: toplevel_meta,
+          jsonapi: toplevel_jsonapi
+        }.reject! {|_,v| v.nil? }
+      end
+
+      # definition:
+      #   JSON Object
+      #
+      # properties:
+      #   version : String
+      #   meta
+      #
+      # description:
+      #   An object describing the server's implementation
+      def toplevel_jsonapi
+        {
+          version: ActiveModelSerializers.config.jsonapi_version,
+          meta: ActiveModelSerializers.config.jsonapi_toplevel_meta
+        }.reject! { |_, v| v.blank? }
+      end
+      alias jsonapi toplevel_jsonapi
+
+      # definition:
+      #   oneOf
+      #     resource
+      #     array of unique items of type 'resource'
+      #     null
+      #
+      # description:
+      #   The document's "primary data" is a representation of the resource or collection of resources
+      #   targeted by a request.
+      #
+      #   Singular: the resource object.
+      #
+      #   Collection: one of an array of resource objects, an array of resource identifier objects, or
+      #   an empty array ([]), for requests that target resource collections.
+      #
+      #   None: null if the request is one that might correspond to a single resource, but doesn't currently.
+      def toplevel_data
+        if serializable_resource.resource?
+          resource
+        elsif serializable_resource.collection?
+          [
+            resource,
+            resource
+          ]
+        else
+          nil
+        end
+      end
+      alias data toplevel_data
+
+      # definition:
+      #   array of unique items of type 'error'
+      def toplevel_errors
+        [
+          error,
+          error
+        ]
+      end
```

Comment:
these should be comments with `# @example` and the implementation is `raise NotImplementedError`